main.go

```diff
@@ -112,17 +128,105 @@ func main() {
+	protos.RegisterProfilesServiceServer(grpcSrv, catalogGrpcServer)
+
+	// handle interrupts
+	shutdownContext, managerCancelFunc := context.WithCancel(context.Background())
+	g, gctx := errgroup.WithContext(context.Background())
+
+	// serve grpc apis
+	setupLog.Info(fmt.Sprintf("starting profiles grpc server at %s", grpcAddr))
+	g.Go(func() error {
+		if err := grpcSrv.Serve(grpcLis); err != nil {
+			setupLog.Error(err, "unable to start grpc api server")
+			return err
+		}
+		return nil
+	})
+
+	// setup grpc-gateway to connect to the grpc server
+	mux := gruntime.NewServeMux()
+	gopts := []grpc.DialOption{grpc.WithInsecure()}
+	if err := protos.RegisterProfilesServiceHandlerFromEndpoint(context.Background(), mux, grpcAddr, gopts); err != nil {
+		setupLog.Error(err, "failed to register service handler from endpoint")
+		os.Exit(1)
+	}
+
+	setupLog.Info(fmt.Sprintf("starting profiles grpc-gateway server at %s", apiAddr))
+	server := &http.Server{Addr: apiAddr, Handler: mux}
+	g.Go(func() error {
+		// ignore server is closing error because the server receives that on graceful shutdown.
+		if err := server.ListenAndServe(); err != nil && !errors.Is(err, http.ErrServerClosed) {
 			setupLog.Error(err, "unable to start profiles api server")
-			os.Exit(1)
+			return err
 		}
-	}()
+		return nil
+	})
 
 	setupLog.Info("starting manager")
-	if err := mgr.Start(ctrl.SetupSignalHandler()); err != nil {
-		setupLog.Error(err, "problem running manager")
+	g.Go(func() error {
+		if err := mgr.Start(shutdownContext); err != nil {
+			setupLog.Error(err, "problem running manager")
+			return err
+		}
+		return nil
+	})
+
+	g.Go(func() error {
+		interruptChannel := make(chan os.Signal, 2)
+		signal.Notify(interruptChannel, os.Interrupt, syscall.SIGTERM)
+
+		select {
+		case <-interruptChannel:
+			done := make(chan struct{})
+			// start the timer for the shutdown sequence
+			go func() {
+				select {
+				case <-done:
+					return
+				case <-time.After(15 * time.Second):
+					setupLog.Error(errors.New("timeout"), "graceful shutdown timed out... forcing shutdown")
+					os.Exit(1)
+				}
+			}()
+			setupLog.Info("received shutdown signal... gracefully terminating servers...")
+			// shutdown the manager
+			managerCancelFunc()
+			// shutdown grpc-gateway server
+			serverTimeoutContext, timeout := context.WithTimeout(context.Background(), 10*time.Second)
+			defer timeout()
+			if err := server.Shutdown(serverTimeoutContext); err != nil {
+				setupLog.Error(err, "Failed to gracefully shutdown server... terminating.")
+			}
+			// shutdown grpc server
+			grpcSrv.GracefulStop()
+			setupLog.Info("all done. Goodbye.")
+		case <-gctx.Done():
+			setupLog.Info("closing signal handler")
+			return gctx.Err()
+		}
+		return nil
+	})
+
+	if err := g.Wait(); err != nil {
+		if errors.Is(err, context.Canceled) {
+			setupLog.Error(err, "context was cancelled")
+		} else {
+			setupLog.Error(err, "server error detected")
+		}
```

Comment:
so much "magic" :cry:

Reply:
Yeah I wish it could be nicer, but this is almost boilerplate graceful shutdown handling core with proper timeouts so it doesn't hang and it gets impossible to kill. I'm tinkering with extracting it, but that kills any defer cancels that we might call here.

Reply:
Actually, my tinkering turned out some pretty awesome stuff! :D 
So stayed tuned for tomorrow, I'll show you the new code. :D Hopefully you'll like that more. :))))))